### PR TITLE
feat: append_file incl. all tests [agentskills]

### PR DESCRIPTION
**Pull request by tobitege**
**_Replacement for discarded first draft:_** #2207 

Adds `append_file` to Agentskills plugin. Includes unit tests and adapted integration tests.

This PR adds a new method **append_file(content: str)** to the agent skills.

In testing OD with some file operations, it is a noticable issue for repeated errors by the model (used Gemini Pro 1.5) that adding content to a file often causes exceptions due to an invalid "start" line number for the edit_file command.

It seems very hard for the model to identify or keep track the total number of lines in a file or it is just not good enough in counting (ask an LLM to count the words of its answer and you know what I mean).
This issue can cause extra cost when it shouldn't.

Such an error then looks like this:
grafik

An example prompt to demonstrate the use of it could be like:
"_Write the numbers 5 to 10 into a new file named test.txt_"
A followup prompt could then just say:
"_Append the numbers 20 to 25 to the same file._"
If it were to use the edit_file command, chances would be high that it would use the wrong line number(s) again.

cc @neubig @li-boxuan 

(blocked by https://github.com/OpenDevin/OpenDevin/pull/2085 as well)

### CODE REVIEW

**Comment by tobitege** [comment]
It'd be cool if one reviewer could trigger the workflow to start just to see how the tests go, but no merging yet.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:367]: You will create a follow-up PR to add `file_name` as part of `edit_file`'s argument anyways, right? Shall we include `file_name` here, rather than rely on the `CURRENT_FILE`?

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:376]: Why not simply append?

File: tests/integration/mock/CodeActAgent/test_edits/prompt_003.log
Comment (li-boxuan) [inline tests/integration/mock/CodeActAgent/test_edits/prompt_003.log:361]: weird, I don't understand why GitHub is showing a diff here... I don't see any changes.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:367]: Correct, in a follow-up... just for the fact that integration tests are a PITA to get right, to be blunt. ;)
That both append_/edit_file then get as 1st param a `file_name` makes total sense.

**Comment by li-boxuan** [comment]
Btw why is this blocked by #2085? I think we could finish this PR, merge it, note it down, and do evaluation afterwards.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:376]: My thought was that it would be a change in behavior, as the other methods error if a file wasn't explicitly opened and/or created first.

File: tests/integration/mock/CodeActAgent/test_edits/prompt_003.log
Comment (tobitege) [reply tests/integration/mock/CodeActAgent/test_edits/prompt_003.log:361]: Probably because I had to restart from scratch with a `main`-level branch and re-apply my changes. Might just be an artifact.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:367]: ok sounds fine to me

**Comment by tobitege** [comment]
> Btw why is this blocked by #2085? I think we could finish this PR, merge it, note it down, and do evaluation afterwards.
> 
> Doing evaluation for every change to prompt might be too costly.

I copied that note about #2085 over, just in case.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:376]: I see. This behavior will be changed soon anyways, right? I mean, you will revert this part in a follow-up PR?

Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:376]: It's the same as in edit_file. I didn't know this was about to be changed.

**Comment by tobitege** [comment]
Ok, great to see that besides Mac, all the unit and integration tests worked. 🎸

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:367]: I am not sure about @neubig 's benchmarking plan. If we decide to do benchmarking before this PR gets merged, then we should probably include the fix to edit_file in the same branch too. Otherwise, we could merge this first, and then edit_file, and maybe a few other PRs, then run evaluation at once.

Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:367]: I decided to work on bringing in the new `file_name` param for both append/edit_file methods.
Core changes and unit tests done. Integration tests next...

File: tests/integration/mock/CodeActAgent/test_edits/prompt_003.log
Comment (enyst) [reply tests/integration/mock/CodeActAgent/test_edits/prompt_003.log:361]: Or maybe some whitespace. If I choose the setting "Hide whitespace" on GitHub, then reload, I no longer see these particular lines in the diff.

Comment (li-boxuan) [reply tests/integration/mock/CodeActAgent/test_edits/prompt_003.log:361]: Probably difference between `\n` and `\r\n`, coz it doesn't seem there's any trailing space.

**Comment by tobitege** [comment]
Commit incoming shortly with `file_name` added as parameter for append-/edit_file methods incl. updated tests...

File: tests/integration/test_agent.py
Comment (tobitege) [inline tests/integration/test_agent.py:29]: @li-boxuan this is one thing that took me hours to find and resolve:
The runtime loads values from the config.toml file, which may have a different workspace_base than what regenerate sets. This resulted in files being read and written in 2 different folders, thus multiple tests failing, like the "bad.txt" one.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: cc @li-boxuan @xingyaoww 
Right now the behavior was that only existing files could be edited and the LLM was supposed to use create_file and/or open_file for that.
I'd just like to confirm before I'd remove that exception for non-existing files, because that would require more changes again to prompts etc.

**Comment by tobitege** [comment]
Also, can someone confirm that this should work without a config.toml present?
`TEST_ONLY=true ONLY_TEST_AGENT="CodeActAgent" ./tests/integration/regenerate.sh`
For me this errors if no config.toml exists.

**Comment by li-boxuan** [comment]
> Also, can someone confirm that this should work without a config.toml present?

Yes. That's exactly how CI is set up.

File: tests/integration/test_agent.py
Comment (li-boxuan) [reply tests/integration/test_agent.py:29]: That's very weird, because the script sets environment variable `WORKSPACE_BASE`, which would override the value in `config.toml`.

Comment (li-boxuan) [reply tests/integration/test_agent.py:29]: I just tried and it worked as expected. I set `workspace_base="./workspace_base_in_toml"` in my `config.toml` and ran the script. I printed out `workspace_base` after L17 and the value indeed was `_test_workspace` (defined in the script).

Comment (li-boxuan) [reply tests/integration/test_agent.py:29]: I am using Mac btw. Not sure if we have some os-dependent behavior somewhere

Comment (tobitege) [reply tests/integration/test_agent.py:29]: > I am using Mac btw. Not sure if we have some os-dependent behavior somewhere

That is what I fear with WSL right now.
Could you try this on your end if the file-related tests work with the config.toml file pointing to some different folder, please?
`TEST_ONLY=true ONLY_TEST_AGENT="CodeActAgent" ./tests/integration/regenerate.sh`
In my case, I have in config.toml:
`workspace_base="/mnt/d/github/workspace"`
which is one folder "below" my OpenDevin repo.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: I am not sure what would be better. If I were you, I would choose whichever is simpler for now, but I don't really have a strong opinion here. Either sounds fine to me...

Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: I'd prefer to leave it as-is for now. Maybe this could be looked closer at after the next benches, what the logfiles show.
Could well be that a change might save trips for some LLM.

File: tests/integration/test_agent.py
Comment (li-boxuan) [reply tests/integration/test_agent.py:29]: I tried, and it worked. I set `workspace_base="/mnt/d/github/workspace"` in my `config.toml` too, which is a non-existent path.

What do you see when you run `echo $WORKSPACE_BASE $WORKSPACE_MOUNT_PATH`? I am guessing it's non-empty, otherwise I cannot explain it.

Comment (tobitege) [reply tests/integration/test_agent.py:29]: I did an echo in regenerate whilst testing and both were set to
`/mnt/d/github/OpenDevin/_test_workspace`
before Run().
Do you think I should remove that block in regenerate.sh again or see side-effects with it?

Comment (li-boxuan) [reply tests/integration/test_agent.py:29]: Nope doesn’t sound necessary. @enyst is there anything you can think of that might affect the order of loading configs?

**Comment by enyst** [comment]
> Also, can someone confirm that this should work without a config.toml present? `TEST_ONLY=true ONLY_TEST_AGENT="CodeActAgent" ./tests/integration/regenerate.sh` For me this errors if no config.toml exists.

It works with no toml for me too. Do you have env vars set, like any of the workspace vars? Any chance you see what error it gives you with no toml?

**Comment by tobitege** [comment]
> Any chance you see what error it gives you with no toml?

My WSL env doesn't have any OpenDevin vars.
I'll give it a try again without toml later today and try to find out more details.

**Comment by tobitege** [comment]
If you guys are ok with the extra check staying - and avoiding potential WSL issues - then I think this PR is good to go. 🚀

File: tests/integration/test_agent.py
Comment (enyst) [reply tests/integration/test_agent.py:29]: There is a thing that happens: config is loaded twice. The second time is when browser_env starts... I have not caught actual errors because of it, but it does mean there's a moment in time when things have started, including sandbox, but config is inconsistent until it's done reloading. Also raises the question what env does that process get.

A shot in the dark here... how about `persist_sandbox` var, do you have it set to true @tobitege ?

Comment (tobitege) [reply tests/integration/test_agent.py:29]: Now that you mention it, I do have that set to true in my config.toml.
I guess I should try with false then? 😁

Comment (enyst) [reply tests/integration/test_agent.py:29]: I don't think it's tested with integration tests? I would set it to false, and clean up both containers and (opendevin) images from docker to be sure.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: Yes, I was already thinking of refactoring the append/edit methods so that both just call one internal method with an extra bool for append, but wanted to do this in followup PR, no problem.

The append is really just to make it easier for the LLM since there were often cases where it tried to start at "end of file + 1" line, got an exception, and had to do it again, thus waisting time+money.

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: Make sense! My inclination is that ideally we can do the refractor, merge it - and when i finish the swe-bench eval, I'll re-run the eval from the start to make sure we are not degrading perf.

**Review by xingyaoww — APPROVED** [review]
LGTM overall! Approve to unblock the PR so @tobitege can work on a refractored version of the `edit_file` - or feel free to do it in this PR as well.

I can run a more comprehensive eval once we fix all the kinks in swe-bench eval and revert when necessary.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: I'd just prefer to get this merged before I do a refactor and in the meantime another patch comes along with changed unit/integration tests. That takes a lot of time to redo.

**Comment by tobitege** [comment]
> I can run a more comprehensive eval once we fix all the kinks in swe-bench eval and revert when necessary.

Ah, excellent, just overlapped with my reply above. Cheers!

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: Approve the PR to unblock you :)

Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:403]: Need a 2nd reviewer 🤣

**Comment by tobitege** [comment]
cc @neubig @li-boxuan  ✉️ kind invitation for review ✏️ 👨 😃

**Comment by tobitege** [comment]
@yufansong please click 🕹️ 😂

File: tests/integration/test_agent.py
Comment (li-boxuan) [reply tests/integration/test_agent.py:29]: So yeah that might very likely be the curse.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:105]: love this!

**Pull request merged**
